src/envoy/http/authn/http_filter_factory.cc

```diff
@@ -0,0 +1,94 @@
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "src/envoy/http/authn/http_filter.h"
+#include "authentication/v1alpha1/policy.pb.h"
+#include "envoy/registry/registry.h"
+#include "google/protobuf/util/json_util.h"
+#include "src/envoy/utils/utils.h"
+
+namespace Envoy {
+namespace Server {
+namespace Configuration {
+
+class AuthnFilterConfig : public NamedHttpFilterConfigFactory,
+                          public Logger::Loggable<Logger::Id::http> {
+ public:
+  HttpFilterFactoryCb createFilterFactory(const Json::Object& config,
+                                          const std::string&,
+                                          FactoryContext& context) override {
+    ENVOY_LOG(debug, "Called AuthnFilterConfig : {}", __func__);
+
+    istio::authentication::v1alpha1::Policy* ptr =
+        new istio::authentication::v1alpha1::Policy();
+    google::protobuf::util::Status status =
+        Utils::ParseJsonMessage(config.asJsonString(), ptr);
+    ENVOY_LOG(debug, "Called AuthnFilterConfig : Utils::ParseJsonMessage()");
+    if (status.ok()) {
+      return createFilter(ptr, context);
+    } else {
+      ENVOY_LOG(debug, "Utils::ParseJsonMessage() return value is NOT ok!");
+      return nullptr;
+    }
+  }
+
+  HttpFilterFactoryCb createFilterFactoryFromProto(
+      const Protobuf::Message& proto_config, const std::string&,
+      FactoryContext& context) override {
+    ENVOY_LOG(debug, "Called AuthnFilterConfig : {}", __func__);
+
+    const istio::authentication::v1alpha1::Policy& policy =
+        dynamic_cast<const istio::authentication::v1alpha1::Policy&>(
+            proto_config);
+
+    return createFilter(&policy, context);
+  }
+
+  ProtobufTypes::MessagePtr createEmptyConfigProto() override {
+    ENVOY_LOG(debug, "Called AuthnFilterConfig : {}", __func__);
+    return ProtobufTypes::MessagePtr{
+        new istio::authentication::v1alpha1::Policy};
+  }
+
+  std::string name() override { return "authN"; }
+
+ private:
+  HttpFilterFactoryCb createFilter(
+      const istio::authentication::v1alpha1::Policy* ptr,
+      FactoryContext& context) {
+    ENVOY_LOG(debug, "Called AuthnFilterConfig : {}", __func__);
+
+    Upstream::ClusterManager& cm = context.clusterManager();
+
+    return [&cm, ptr](Http::FilterChainFactoryCallbacks& callbacks) -> void {
+      callbacks.addStreamDecoderFilter(
+          std::make_shared<Http::AuthenticationFilter>(
```

Comment:
share object should be created outside of lambda, otherwise it is NOT shared.

Reply:
Done. Has been revised based on the discussion.